---
title: "ggtree: a phylogenetic tree viewer for different types of tree annotations"
author: "\\

	Guangchuang Yu\\

	<guangchuangyu@gmail.com>\\

        School of Public Health, The University of Hong Kong"
date: "`r Sys.Date()`"
bibliography: ggtree.bib
csl: nature.csl
output: 
  BiocStyle::html_document:
    toc: true
  BiocStyle::pdf_document:
    toc: true
vignette: >
  %\VignetteIndexEntry{ggtree: phylogenetic tree viewer and annotator}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---

```{r style, echo=FALSE, results="hide", message=FALSE}
BiocStyle::markdown()
knitr::opts_chunk$set(tidy = FALSE,
		   message = FALSE)
```


```{r echo=FALSE, results="hide", message=FALSE}
library("ape")
library("colorspace")
library("ggplot2")
library("ggtree")
library("Biostrings")
library("gridExtra")
```



> You can't even begin to understand biology, you can't understand life, unless you understand what it's all there for, how it arose - and that means evolution.
> --- Richard Dawkins


# Introduction
This project arose from my needs to annotate nucleotide substitutions in the phylogenetic tree, and I found that there is no tree visualization software can do this easily. Existing tree viewers are designed for displaying phylogenetic tree, but not annotating it. Although some tree viewers can displaying bootstrap values in the tree, it is hard/impossible to display other information in the tree. My first solution for displaying nucleotide substituitions in the tree is to add this information in the node/tip names and use traditional tree viewer to show it. I displayed the information in the tree successfully, but I believe this indirect approach is inefficient.

In the old day, phylogenetic tree is often small. At that time, as we almost didn't have a need to annotate a tree; displaying the evolution relationships is mostly enough. Nowadays, we can obtain a lot of data from different experiments, and we want to associate our data, for instance antigenic change, with the evolution relationship. Visualizing these associations in the phylogenetic tree can help us to identify evolution patterns. I believe we need a next generation tree viewer that can view a phylogenetic tree easily as we did with classical software and support adding annotation data in a layer above the tree. This is the objective of developing the `r Githubpkg("GuangchuangYu/ggtree")`. Common tasks of annotating a phylogenetic tree should be easy and complicated tasks can be possible to achieve by adding multiple layers of annotation.

The `r Githubpkg("GuangchuangYu/ggtree")` is designed by extending the `r CRANpkg("ggplot2")`[@wickham_ggplot2_2009] package. It is based on the grammar of graphics and takes all the good parts of `r CRANpkg("ggplot2")`. There are other R packages that implement tree viewer using `r CRANpkg("ggplot2")`, including `r CRANpkg("OutbreakTools")`, `r Biocpkg("phyloseq")`[@mcmurdie_phyloseq_2013] and `r Githubpkg("gjuggler/ggphylo")`; they mostly create complex tree view functions for their specific needs. The `r Githubpkg("GuangchuangYu/ggtree")` is different to them by allowing general flexibilities of annotating phylogenetic tree with diverse types of user inputs. 

# Tree visualization
## viewing tree with `ggtree`
`r Githubpkg("GuangchuangYu/ggtree")` extend _`ggplot`_ to support viewing phylogenetic tree. It implements _`geom_tree`_ layer for displaying phylogenetic trees, as shown below:
```{r fig.width=3, fig.height=3, fig.align="center"}
nwk <- system.file("extdata", "sample.nwk", package="ggtree")
x <- readLines(nwk)
cat(substring(x, 1, 56), "\n", substring(x, 57), "\n")

library("ggplot2")
library("ggtree")

tree <- read.tree(nwk)
ggplot(tree, aes(x, y)) + geom_tree() + theme_tree() + xlab("") + ylab("")
```

This example tree was obtained from Chapter 34 of _Inferring Phylogenies_[@felsenstein_inferring_2003 pp.573]. The function, _`ggtree`_, was implemented as a short cut to visualize a tree, and it works exactly the same as shown above.

_`ggtree`_ takes all the advantages of _`ggplot2`_. For example, we can change the color, size and type of the lines as we do with _`ggplot2`_.
```{r fig.width=3, fig.height=3, fig.align="center"}
ggtree(tree, color="steelblue", size=0.5, linetype="dotted")
```

By default, the tree is viewing in ladderize form, user can set the parameter _`ladderize = FALSE`_ to disable it.
```{r fig.width=3, fig.height=3, fig.align="center"}
ggtree(tree, ladderize=FALSE)
```

The _`branch.length`_ is used to scale the edge, user can set the parameter _`branch.length = "none"`_ to only viewing the tree topology.

```{r fig.width=3, fig.height=3, fig.align="center"}
ggtree(tree, branch.length="none")
```

## layout

Currently, _`ggtree`_ supports several layout, including:

 + `phylogram` (by default)
 + `cladogram`
 + `dendrogram` 
 + `fan`
 + `unrooted`. 

Unrooted layout was implemented by the _`equal-angle algorithm`_ that described in _Inferring Phylogenies_[@felsenstein_inferring_2003 pp.578-580].

```{r fig.width=6, fig.height=9, fig.align="center"}
library("gridExtra")
grid.arrange(ggtree(tree) + ggtitle("phylogram layout"),
	     ggtree(tree, layout="dendrogram") + ggtitle("dendrogram layout"),
	     ggtree(tree, layout="cladogram") + ggtitle("cladogram layout"),
	     ggtree(tree, layout="cladogram", branch.length="none") + 
	            scale_x_reverse()+coord_flip() + ggtitle("cladogram layout"),
	     ggtree(tree, layout="fan") + ggtitle("fan layout"),
	     ggtree(tree, layout="unrooted") + ggtitle("unrooted layout"),
	     ncol=2)
```


## support multiple phylogenetic classes

`r Githubpkg("GuangchuangYu/ggtree")` defined several S4 classes to store phylogenetic object and its associated annotation, including:

+ `jplace`
+ `palm_rst`
+ `codeml_mlc`
+ `codeml`
+ `hyphy`
+ `beast`

In addition, it also supports _`phylo`_ (defined by `r CRANpkg("ape")`[@paradis_ape_2004]), and _`phylo4`_ (defined by `r CRANpkg("phylobase")`)

User can use _`ggtree(object)`_ command to view the phylogenetic tree directly, and annotation data stored in these objects can be added as demonstrated in __[`Tree annotation`](#tree-annotation)__ session.
 
## display evolution distance

To show evolution distance, we can use `theme_tree2()` or `ggtree(showDistance=TRUE)`

```{r fig.width=3, fig.height=3, fig.align="center"}
ggtree(tree) + theme_tree2()
```

Another way is to show the edge length of the tree. Besides, the scale of branch length can be specify via _`scale_x_continuous()`_. 
```{r fig.width=3, fig.height=3, warning=FALSE, fig.align="center"}
ggtree(tree, showDistance=TRUE) +
      geom_text(aes(label=branch.length, x=branch), size = 3, 
      		vjust=-0.5, color="#F06C45") +
      scale_x_continuous(breaks=seq(0, 60, 5))
```

User can also use _`geom_segment`_ and _`geom_text`_ to specify the position and length of distance legend:
```{r fig.width=3, fig.height=3, warning=FALSE, fig.align="center"}
ggtree(tree) +
       geom_segment(x=0, y=12, xend=10, yend=12) +
              geom_text(x=5, y=12.5, label=5, size=4)
```

## display nodes/tips

Show all the internal nodes and tips in the tree can be done by adding a layer of points using _`geom_point`_.

```{r fig.width=3, fig.height=3, fig.align="center"}
ggtree(tree)+geom_point(aes(shape=isTip, color=isTip), size=3)
```
 
And of course, we can separate nodes and tips by using _`subset`_.
```{r fig.width=3, fig.height=3, fig.align="center"}
p <- ggtree(tree) + geom_point(subset=.(!isTip), 
     		       	       color="#b5e521", alpha=1/4, size=10)
p + geom_point(color="#FDAC4F", shape=8, size=3, subset=.(isTip))
```

## display labels

Users can use _`geom_text`_ to display the node/tip labels:

```{r fig.width=3, fig.height=3, warning=FALSE, fig.align="center"}
p + geom_text(aes(label=label), size=3, color="purple", hjust=-0.3)
```

For _`fan`_ and _`unrooted`_ layout, `r Githubpkg("GuangchuangYu/ggtree")` supports rotating node labels according to the angles of the branches.
```{r fig.width=6, fig.height=6, warning=FALSE, fig.align="center"}
ggtree(tree, layout="fan") + geom_text(aes(label=label, angle=angle), size=3, color="purple", vjust=-0.3)
```


By default, the positions are based on the node positions, we can change them to based on the middle of the branch/edge.

```{r fig.width=5, fig.height=3, warning=FALSE, fig.align="center"}
p + geom_text(aes(x=branch, label=label), size=3, color="purple", vjust=-0.3)
```

Based on the middle of branches is very useful when annotating transition from parent node to child node.

## theme
`theme_tree()` defined a totally blank canvas, while _`theme_tree2()`_ add phylogenetic distance legend. These two themes all accept a parameter of _`bgcolor`_ that defined the background color.

```{r fig.width=6, fig.height=3, fig.align="center"}
grid.arrange(
	ggtree(rtree(30), color="red") + theme_tree("steelblue"),
	ggtree(rtree(20), color="white") + theme_tree("black"),
	ncol=2)
```

## update tree viewing with a new tree
In the [display nodes/tips](#display-nodestips) section, we have a _`p`_ object that stored the tree viewing of 13 tips and internal nodes highlighted with specific colored big dots. If you want to applied this pattern (we can imaging a more complex one) to a new tree, we don't need to build the tree step by step. `r Githubpkg("GuangchuangYu/ggtree")` provides an operator, _`%<%`_, for applying the visualization pattern to a new tree.

For example, the pattern in the _`p`_ object will be applied to a new tree with 50 tips as shown below:
```{r fig.width=3, fig.height=3, fig.align="center"}
p %<% rtree(50)
```

Another example can be found in [`CODEML`](#codeml) session.


# Tree annotation
In `r Githubpkg("GuangchuangYu/ggtree")`, we implemented several functions to parse the output from [PAML](http://abacus.gene.ucl.ac.uk/software/paml.html)[@yang_paml_2007], [HYPHY](http://hyphy.org/w/index.php/Main_Page)[@pond_hyphy_2005], [EPA](http://sco.h-its.org/exelixis/web/software/epa/index.html)[@berger_EPA_2011], [PPLACER](http://matsen.fhcrc.org/pplacer/)[@matsen_pplacer_2010] and [BEAST](http://beast2.org/)[@bouckaert_beast_2014] and defined several classes to store phylogenetic object and associated annotation.

Classes include:

+ _`palm_rst`_ for _`rst`_ file obtained by [PAML](http://abacus.gene.ucl.ac.uk/software/paml.html)[@yang_paml_2007], including _`BASEML`_ and _`CODEML`_.
+ _`codeml_mlc`_ for _`mlc`_ file obtained by _`CODEML`_.
+ _`codeml`_ for interpreting _`rst`_ and _`mlc`_ files obtained by _`CODEML`_.
+ _`hyphy`_ for [HYPHY](http://hyphy.org/w/index.php/Main_Page)[@pond_hyphy_2005] output.
+ _`jplace`_ for [EPA](http://sco.h-its.org/exelixis/web/software/epa/index.html)[@berger_EPA_2011] and [PPLACER](http://matsen.fhcrc.org/pplacer/)[@matsen_pplacer_2010] output.
+ _`beast`_ for [BEAST](http://beast2.org/)[@bouckaert_beast_2014]

_`jplace`_ class is also designed to store user specific annotation data, and serves as a standard format for tree annotation within the `r Githubpkg("GuangchuangYu/ggtree")` package. Please refer to the __[`jplace file format`](#jplace-file-format)__ session.

For each classes, we defined _`read.className`_ to parse input file and output a corresponding object, _`get.fields`_ method to get the annotation features available in the object, access methods to get these features, and _`plot`_ methods for quickly viewing these annotation features.


## annotating tree with BEAST output
```{r}
file <- system.file("extdata/BEAST", "beast_mcc.tree", package="ggtree")
beast <- read.beast(file)
beast
```
Since _`%`_ is not a valid character in _`names`_, all the feature names that contain _`x%`_ will convert to _`0.x`_. For example, _`length_95%_HPD`_ will be changed to _`length_0.95_HPD`_.

```{r fig.width=8, width=60, warning=FALSE, fig.align="center"}
plot(beast, annotation="length_0.95_HPD", branch.length="none") + theme_tree()
```

User can round the digits by setting the parameter _`ndigits`_. The default value is 2.

```{r fig.width=8, width=60, warning=FALSE, fig.align="center"}
plot(beast, annotation="height", ndigits=3, annotation.color="red")
```



## annotating tree with PAML output

### BASEML

#### _`rst`_ file\



_`rst`_ file from _`baseml`_ is similar to _`codeml`_ output. The only difference is the space in the sequences. For _`baseml`_, each ten bases are separated by one space, while for _`codeml`_, each three bases (triplet) are separated by one space. We defined a _`read.paml_rst`_ to parse _`rst`_ file. It supports _`baseml`_ and _`codeml`_ output. The information will be stored in _`paml_rst`_ S4 object.

```{r}
rstfile <- system.file("extdata/PAML_Baseml", "rst", package="ggtree")
tipfas <- system.file("extdata", "pa.fas", package="ggtree")
rst <- read.paml_rst(rstfile, tipfas)
rst
```

The function _`read.paml_rst`_ can accept only one parameter, _`rstfile`_, and the output can be used to view the phylogeny. But if we want to view the substitution annotation, we should provide tip sequence fasta file to parameter _`tip.fasfile`_, since _`rstfile`_ only contain inferred ancestral sequences.

_`mlb`_ file of _`baseml`_ output do contain tip sequences, and `r Githubpkg("GuangchuangYu/ggtree")` provides another function _`read.baseml`_ to parse _`rstfile`_ and _`mlbfile`_ simultaneously.

So for viewing substitution annotation, we can use _`read.paml_rst`_ with _`rstfile`_ and _`tip.fasfile`_ provided or use _`read.baseml`_ with _`rstfile`_ and _`mlbfile`_ provided.

```{r}
mlbfile <- system.file("extdata/PAML_Baseml", "mlb", package="ggtree")
baseml <- read.baseml(rstfile, mlbfile)
baseml
```

```{r fig.width=10, fig.height=8, width=60, warning=FALSE}
p <- plot(rst, annotation="marginal_AA_subs", annotation.color="steelblue")
print(p)
```

The following command will generate the same figure.
```{r eval=FALSE}
plot(baseml, annotation="marginal_AA_subs", annotation.color="steelblue")
```

### CODEML

#### rst file\



_`rst`_ file from _`CODEML`_ is similar to _`BASEML`_, and also parsed by _`read.paml_rst`_ function. The _`plot`_ method works also in the same way.

If you remember the _`%<%`_ operator introduced in [`update tree viewing with a new tree`](#update-tree-viewing-with-a-new-tree) session, you can use it to update a tree view with a new object.

In last session, we use _`rstfile`_ of _`BASEML`_ to build a tree view with amino acid substitution annotated. The following example use another _`rstfile`_ from _`CODEML`_ to update the tree view.
```{r fig.width=8, width=60, warning=FALSE, fig.align="center"}
rstfile <- system.file("extdata/PAML_Codeml", "rst", package="ggtree")
rst <- read.paml_rst(rstfile, tipfas)
p %<% rst
```

You can found that these two figures have different evolution distances, and substitutions inferred from _`BASEML`_ and _`CODEML`_ are slighly different.


#### mlc file\


_`mlcfile`_ contains _`dN/dS`_ estimation.

```{r}
mlcfile <- system.file("extdata/PAML_Codeml", "mlc", package="ggtree")
mlc <- read.codeml_mlc(mlcfile)
mlc
```

Please aware that _`/`_ and _`*`_ are not valid characters in _`names`_, they were changed to _`_vs_`_ and _`_x_`_ respectively.

So _`dN_vs_dS`_ is _`dN/dS`_, _`N_x_dN`_ is _`N*dN`_, and _`S_x_dS`_ is _`S*dS`_.


```{r fig.width=8, width=60, warning=FALSE, fig.align="center"}
plot(mlc, branch.length="branch.length", annotation="dN_vs_dS", annotation.color="blue", ndigits=3)
```

The paramter _`branch.length`_ can be one of available annotations:
```{r}
get.fields(mlc)
```

For example, if we set _`branch.length`_ to _`dN_vs_dS`_, it will plot the $\omega$ (_`dN/dS`_) tree:
```{r fig.width=8, width=60, warning=FALSE, fig.align="center"}
plot(mlc, branch.length="dN_vs_dS", annotation="dN_vs_dS", ndigits=3)
```

We can also plot the _`dN`_ or _`dS`_ tree and others. The parameter _`annotation`_ can also be one of the available annotations.


#### _`CODEML`_ output: rst and mlc files\


We annotate the tree with information presented in _`rstfile`_ and _`mlcfile`_ separately as demonstrated in previous sessions.

We can also use both of them and it's highly recommended. User don't need to provide tip sequences, as it's already available in _`mlcfile`_. All the features in both files are available for annotation.


```{r}
ml <- read.codeml(rstfile, mlcfile)
ml
```

So we can annotate _`dN/dS`_ with the tree in _`rstfile`_ and amino acid substitutions with the tree in _`mlcfile`_.

```{r fig.width=12, fig.height=8, width=60, warning=FALSE, fig.align="center"}
plot(ml, branch.length="rst.branch.length", annotation="dN_vs_dS")
plot(ml, branch.length="mlc.branch.length", annotation="marginal_AA_subs")
plot(ml, branch.length="dN", annotation="joint_AA_subs", annotation.color="darkgreen")
```

## annotating tree with HYPHY output
```{r}
nwk <- system.file("extdata/HYPHY", "labelledtree.tree", package="ggtree")
ancseq <- system.file("extdata/HYPHY", "ancseq.nex", package="ggtree")
hy <- read.hyphy(nwk, ancseq, tipfas)
hy
```

```{r fig.width=12, fig.height=10, width=60, warning=FALSE, fig.align="center"}
plot(hy, annotation="AA_subs")
```


## annotating tree with EPA and PPLACER output
[EPA](http://sco.h-its.org/exelixis/web/software/epa/index.html)[@berger_EPA_2011] and [PPLACER](http://matsen.fhcrc.org/pplacer/)[@matsen_pplacer_2010] have common output file format, `jplace`. 

```{r}
jpf <- system.file("extdata/sample.jplace",  package="ggtree")
jp <- read.jplace(jpf)
print(jp)
```

In `r Githubpkg("GuangchuangYu/ggtree")`, we provide _`get.placements`_ method to access the placement.

```{r}
## get only best hit
get.placements(jp, by="best")
## get all placement
get.placements(jp, by="all")
```

This is only a tiny sample file. In reality, [EPA](http://sco.h-its.org/exelixis/web/software/epa/index.html) and [PPLACER](http://matsen.fhcrc.org/pplacer/) may place thousands of short reads on a reference tree.


We may, for example, count the number of placement and annotate this information in the tree. We do not provide a _`plot`_ method for _`jplace`_ object, since we use this file format as a standard annotation format in `r Githubpkg("GuangchuangYu/ggtree")` package and have no assumption of information it may stored. Please refer to [jplace file format](#jplace-file-format) session.


## annotating tree using _`ggplot2`_ layers
We implemented several _`plot`_ methods for easily viewing annotation data. Users are not restricted to _`plot`_ methods provided. They can use _`geom_text`_ to add annotation layer. All annotation data are visible to _`ggplot2`_.

In the following example, we use the _`codeml`_ object to visualize the $\omega$ (_`dN/dS`_) tree, and annotate the tree with _`dN`_ and _`dS`_.
```{r fig.width=12, fig.height=10, width=60, warning=FALSE, fig.align="center"}
ggtree(ml, branch.length="dN_vs_dS") + 
	   geom_text(aes(x=branch, label=dN), 
	   	     size=3, vjust=-0.5, color="red") +
           geom_text(aes(x=branch, label=dS),
	             size=3, vjust=1.2, color="darkgreen")
```


## user specific annotation
We provides several functions to parse and store information from common software output, and corresponding _`plot`_ methods for visualizing annotation in the tree.

Here, we would like to demonstrate how to inject user specific annotation data in a tree.

Suppose we have the following data that associated with the tree and would like to attach the data in the tree.

```{r}
nwk <- system.file("extdata", "sample.nwk", package="ggtree")
tree <- read.tree(nwk)
p <- ggtree(tree)

dd <- data.frame(taxa=LETTERS[1:13], 
      		 place=c(rep("GZ", 5), rep("HK", 3), rep("CZ", 4), NA),
                 value=round(abs(rnorm(13, mean=70, sd=10)), digits=1))
## you don't need to order the data
## data was reshuffled just for demonstration
dd <- dd[sample(1:13, 13), ]
row.names(dd) <- NULL
```
```{r eval=FALSE}
print(dd)
```

```{r echo=FALSE, results='asis'}
knitr::kable(dd)
```

We can imaging that the _`place`_ column is the place we isolated the species and _`value`_ column stored numerical values for example bootstrap values.

We have shown using the operator, _`%<%`_, to update a tree view with a new tree. Here, we will introduce another operator, _`%<+%`_, that attaches annotation data to a tree view. The only requirement of the input data is that its first column should be matched with the node/tip labels of the tree.

After attaching the annotation data to the tree by _`%<+%`_, all the columns in the data are visible to _`ggplot2`_. As an example, here we attach the above annotation data to the tree view, _`p`_, and add a layer that showing the tip labels and colored them by the isolation site stored in _`place`_ column.

```{r fig.width=6, fig.height=5, warning=FALSE, fig.align="center"}
p <- p %<+% dd + geom_text(aes(color=place, label=label), hjust=-0.5) + 
       geom_point(aes(size=value, shape=place, color=place), alpha=0.25, subset=.(isTip))
print(p)
```

Once the data was attached, it is always attached. So we can add another layer to display the isolation sites easily.
```{r fig.width=6, fig.height=5, warning=FALSE, fig.align="center"}
p <- p + geom_text(aes(color=place, label=place), hjust=1, vjust=-0.4, size=3)
print(p)
```

And another layer showing numerical values:
```{r fig.width=6, fig.height=5, warning=FALSE, fig.align="center"}
p <- p + geom_text(aes(color=place, label=value), hjust=1, vjust=1.4, size=3)
print(p)
```

## `jplace` file format
The _`jplace`_ file format was defined by Masten[@matsen_format_2012] for phylogenetic placements. We employed this file format to store phylogenetic tree and user specific annotation data. Suppose we have a tree, and the associated data as shown below:
```{r}
tree <- system.file("extdata", "pa.nwk", package="ggtree")
data <- read.csv(system.file("extdata", "pa_subs.csv", package="ggtree"), stringsAsFactor=FALSE)
print(tree)
head(data)
```

The _`data`_ contains amino acid substitutions from parent node to child node and GC contents of each node. We can annotate the tree as demonstrated in [user specific annotation](#user-specific-annotation) session.

`r Githubpkg("GuangchuangYu/ggtree")` provides a function, _`write.jplace`_, to combine a tree and an associated data and store them to a single _`jplace`_ file.
```{r}
outfile <- tempfile()
write.jplace(tree, data, outfile)
```

Then _`read.jplace`_ function was designed to read the _`jplace`_ file and store the information to a _`jplace`_ object.
```{r}
jp <- read.jplace(outfile)
print(jp)
```

Now we know the _`jp`_ object stored the tree and the associated amino acid substitution and GC content information, we can view the tree and display the associated annotation data on it directly by _`ggtree`_.

```{r fig.width=12, fig.height=12, warning=FALSE, fig.align="center"}
ggtree(jp, showDistance=TRUE) + 
	   geom_text(aes(x=branch, label=subs), color="purple", vjust=-1, size=3) + 
	   geom_text(aes(label=gc), color="steelblue", hjust=-.6, size=3) +
	   geom_text(aes(label=label), hjust=-.5)
```

## visualize tree and associated matrix
```{r}
seqs <- get.tipseq(ml)
library(Biostrings)
x <- DNAStringSet(seqs)

dd <- alphabetFrequency(x)[, 1:4]
dd <- dd/rowSums(dd)
row.names(dd) <- names(seqs)
```

```{r eval=FALSE}
head(dd)
```

```{r echo=FALSE}
knitr::kable(head(dd))
```

```{r fig.width=12, fig.height=6, fig.align="center"}
p <- ggtree(ml)
gplot(p, dd, low="green", high="red", widths=c(.7, .3))
```

Of course, we can use an annotated tree.

```{r fig.width=12, fig.height=6, fig.align="center", warning=FALSE}
p2 <- ggtree(ml, branch.length="none") + 
          geom_text(aes(x=branch, label=dN_vs_dS, color=dN_vs_dS), 
                    vjust=-.5, size=4) +
             scale_color_gradient(low="darkgreen", high="red")	  
gplot(p2, dd, low="green", high="red", widths=c(.7, .3))
```

## zoom on a portion of tree

`r Githubpkg("GuangchuangYu/ggtree")` provides _`gzoom`_ function that similar to _`zoom`_ function provided in `r CRANpkg("ape")`. This function plots simultaneously a whole phylogenetic tree and a portion of it. It aims at exploring very large trees.

```{r fig.width=18, fig.height=10, fig.align="center"}
library("ape")
data(chiroptera)
library("ggtree")
gzoom(chiroptera, grep("Plecotus", chiroptera$tip.label))
```

## highlight clades

`r Githubpkg("GuangchuangYu/ggtree")` implements _`geom_hilight`_ layer, that accepts tree object and internal node number and add a layer of rectangle to highlight the selected clade.

```{r fig.width=5, fig.height=5, fig.align="center", warning=FALSE}
nwk <- system.file("extdata", "sample.nwk", package="ggtree")
tree <- read.tree(nwk)

ggtree(tree) + geom_text(aes(label=node))
```

User can use _`geom_text`_ to display all the node numbers, and select interesting clade to highlight.

```{r fig.width=5, fig.height=5, fig.align="center", warning=FALSE}
ggtree(tree) + geom_hilight(tree, node=21, fill="steelblue", alpha=.6) +
     geom_hilight(tree, node=17, fill="darkgreen", alpha=.6)
```


```{r fig.width=5, fig.height=5, fig.align="center", warning=FALSE}
ggtree(tree, layout="fan") + geom_hilight(tree, node=21, fill="steelblue", alpha=.6) +
     geom_hilight(tree, node=17, fill="darkgreen", alpha=.6)
```

## scale color based on numerical attribute

`r Githubpkg("GuangchuangYu/ggtree")` provides _`scale_color`_ to calculate colors based on numerical attribute of the tree. The output can be used to color lines and annotation text.
```{r fig.width=12, fig.height=6, fig.align="center", warning=FALSE}
cols = scale_color(ml, by="dN")
ggtree(ml, color=cols) + geom_text(aes(label=label), color=cols, hjust=.25)
```

_`scale_color`_ will auto determine the color. User can also provide parameter _`low`_ and _`high`_ to speicify the color scale.

```{r fig.width=12, fig.height=6, fig.align="center", warning=FALSE}
cols = scale_color(ml, by="dN", low="green", high="red")
ggtree(ml, color=cols) + geom_text(aes(label=label), color=cols, hjust=.25)
```

## group OTUs

`r Githubpkg("GuangchuangYu/ggtree")` provides _`groupOTU`_ function to group tips and all their related ancestors. It return a cluster index of each line segment in the tree view.

```{r}
cluster_index <- groupOTU(tree, focus=c("A", "B", "C", "D", "E"))
cluster_index
```
In the _`cluster_index`_, **1** represent the cluster that not selected, while other number represent the corresponding selected group(s).

```{r fig.width=5, fig.height=5, fig.align="center", warning=FALSE}
ggtree(tree, color=c("black", "red")[cluster_index])
```

_`groupOTU`_ can also input a list of tip groups.

```{r fig.width=5, fig.height=5, fig.align="center", warning=FALSE}
cls <- list(c1=c("A", "B", "C", "D", "E"),
            c2=c("F", "G", "H"),
            c3=c("L", "K", "I", "J"),
            c4="M")

cls_ind <- groupOTU(tree, cls)
library("colorspace")
cols <- rainbow_hcl(4)
cols <- c("black", cols)
ggtree(tree, color=cols[cls_ind]) + geom_text(aes(label=label), color=cols[cls_ind], hjust=-.25)
```

We can change the linetype either:
```{r fig.width=5, fig.height=5, fig.align="center", warning=FALSE}
linetype <- c("solid", "dotted", "dashed", "dotdash", "longdash")
ggtree(tree, color=cols[cls_ind], linetype=linetype[cls_ind]) +
      geom_text(aes(label=label), color=cols[cls_ind], hjust=-.25)
```

And also size:
```{r fig.width=5, fig.height=5, fig.align="center", warning=FALSE}
size <- 1:5
ggtree(tree, color=cols[cls_ind], linetype=linetype[cls_ind], size=size[cls_ind]) +
      geom_text(aes(label=label), color=cols[cls_ind], hjust=-.25)
```

All the tree classes defined in `r Githubpkg("GuangchuangYu/ggtree")`, including _`beast`_, _`paml_rst`_, _`codeml_mlc`_, _`codeml`_, _`hyphy`_ and _`jplace`_ are all supported.


### iris example

In this example, we first build a tree based on the iris data, then grouping the tree based on different spacies.
```{r fig.width=20, fig.height=20, fig.align="center", warning=FALSE}
data(iris)
rn <- paste0(iris[,5], "_", 1:150)
rownames(iris) <- rn
d_iris <- dist(iris[,-5], method="man")

tree_iris <- bionj(d_iris)
cls_ind <- groupOTU(tree_iris, list(setosa    = rn[1:50],
				versicolor    = rn[51:100],
				virginica_145 = rn[101:150]))
cols <- rainbow_hcl(4)
ggtree(tree_iris, color=cols[cls_ind]) +
     geom_text(aes(label=label), color=cols[cls_ind], hjust=-.1)
```

<!--
By adding a layer of internal node number, we can easily extract tip labels of a particular clade by the _`get.offspring.tip`_ function.
```{r}
ggtree(tree_iris) + geom_text(aes(label=node))
cl1 <- get.offspring.tip(tree_iris, 242)
cl2 <- get.offspring.tip(tree_iris, 152) 
cl2 <- cl2[!cl2 %in% cl1]
cl3 <- get.offspring.tip(tree_iris, 158)
cl4 <- get.offspring.tip(tree_iris, 157)

cls_ind <- groupOTU(tree_iris, list(cl1, cl2, cl3, cl4))
```

```{r fig.width=20, fig.height=24, fig.align="center", warning=FALSE}
cols <- rainbow_hcl(3)
cols <- c("black", cols, cols[2])
species <- data.frame(otu=row.names(iris), species=iris[,5])

ggtree(tree_iris, color=cols[cls_ind]) %<+% species + 
     geom_text(aes(label=label, color=species), hjust=-0.1) +
         scale_color_manual(values=cols[2:4])
```
-->


This example demonstrates how the separation of the _`bionj`_ is very good with the _`setosa`_ species, but misses in labeling several _`versicolor`_ and _`virginica`_ species.

## phylopic

[PhyloPic](http://phylopic.org/) is a database that stores reusable silhouette images of organisms. `r Githubpkg("GuangchuangYu/ggtree")` supports downloading images from [PhyloPic](http://phylopic.org/) and annotating phylogenetic tree with the downloaded images.

```{r fig.width=5, fig.height=5, fig.align="center", warning=FALSE}
cols <- rainbow_hcl(2)
x <- get.phylopic("d3563b54-780f-4711-a49a-7ea051e9dacc", size=512, color=cols[1], alpha=.3)
y <- get.phylopic("3318565d-7531-4a8c-8b84-2e87dfae36ab", size=512, color=cols[2], alpha=.8)
ggtree(tree) + annotation_custom(xmin=-Inf, ymin=-Inf, xmax=Inf, ymax=Inf, x) +
       annotation_custom(xmin=15, ymin=2, xmax=25, ymax=4, y) 
```



# Session info
Here is the output of `sessionInfo()` on the system on which this document was compiled:
```{r echo=FALSE}
sessionInfo()
```


# References